### PR TITLE
Add configurable PRE and POST execution hooks.

### PR DESCRIPTION
**Pull request by DnlLrssn**
**Checklist**

- [x] tests added
- [x] commits conform to the [Commit message format](https://github.com/qlik-oss/gopherciser/blob/master/.github/CONTRIBUTING.md#commit)
- [x] documentation updated

**Squash commit message**

* Moves synced templates to their own package
* Adds pre and post execution hooks which can be used to send requests with data before and after test is executed.
* Add floats NearlyEqual and NearlyEqualEpsilon helper functions.

**Info**

Optional informative text (not to be included in commit message) relevant to reviewers.

### CODE REVIEW

File: config/hooks.go [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+	req, err := http.NewRequestWithContext(ctx, strings.ToUpper(httpMethodEnum.StringDefault(int(hook.Method), http.MethodPost)), hook.Url, buf)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	req.Header.Set("Content-Type", hook.ContentType)
+	for k, v := range headers {
+		req.Header.Set(k, v)
+	}
+
+	// TODO try avoid creating client here
```
Comment (SeBBBe) [inline config/hooks.go:302]: It will only be done once pre and once post test though, right? So should be a small overhead

**Review by SeBBBe — APPROVED** [review]
Will be some work if we want to implement it in the GUI but otherwise looks great! I verified that opening and saving a script without affecting the hook settings work though so we're all good from that point of view.

File: config/hooks.go (hunk 1 above)
Comment (DnlLrssn) [reply config/hooks.go:302]: Correct. But if we need to change any of the settings at some point we might forget to change this one too.

**Comment by DnlLrssn** [comment]
> Will be some work if we want to implement it in the GUI 

ok any part in particular that doesn't work out of the box?

**Comment by SeBBBe** [comment]
> > Will be some work if we want to implement it in the GUI
> 
> ok any part in particular that doesn't work out of the box?

All the root level parts require some special handling to appear so can't say for sure. All the components for strings, arrays etc are automatically reused but the base level expander needs to be manually inserted in one of the panes.

The part I worry about the most would be the extractors slice but I can have a closer look and see what happens.

**Comment by SeBBBe** [comment]
Screenshot 2021-05-06 at 13 21 01

**Comment by SeBBBe** [comment]
It mostly works.
* Extractors can't be created from GUI
* There's no distinction between which is the pre and post hook:

Screenshot 2021-05-06 at 13 22 03

**Comment by SeBBBe** [comment]
If I create a new script and try to save it immediately, I get this validation error:

`config.HookCore.Content: Method: unmarshalerDecoder: Failed to unmarshal httpMethodEnum: Unknown enum<>: Key<> not found, error found in #10 byte of ...|method":"","payload"|..., bigger context ...|{"contenttype":"","extractors":[],"method":"","payload":"","respcodes":[],"url":""}|...`

**Comment by SeBBBe** [comment]
Seems like creating extractors fails because this:

sliceElemType := reflect.TypeOf(input).Elem()
sampleValue := reflect.New(sliceElemType) // Instantiate new member of slice element type
sample := sampleValue.Elem().Interface()
inputType := reflect.TypeOf(input)

Ends up with inputType == nil

File: helpers/floats.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+// NearlyEqual float points covering some edge cases as per https://floating-point-gui.de/errors/comparison/#look-out-for-edge-cases
+func NearlyEqualEpsilon(a, b float64, epsilon float64) bool {
+	absA := math.Abs(a)
+	absB := math.Abs(b)
+	diff := math.Abs(a - b)
+
+	if a == b { // shortcut, handles infinities
+		return true
+	} else if a == 0 || b == 0 || diff < FloatMinNormal {
```
Comment (atluq) [inline helpers/floats.go:25]: `diff` shall be `absA + absB`

**Review by atluq — COMMENTED** [review]
Looks good!

Just a typo in the floating point comparison implementation.

File: helpers/floats.go (hunk 2 above)
Comment (DnlLrssn) [reply helpers/floats.go:25]: hmm interesting, actually converted the code the shows up in duckduckgo which is from stackoverflow (but wanted to link to original author), comparing what's on the site and what shows there is different, must have been updated
image

Comment (DnlLrssn) [reply helpers/floats.go:25]: good find. will update.

**Review by SeBBBe — APPROVED** [review]

**Comment by SeBBBe** [comment]
After some adjustments by Daniel and to QSEST by me, it's now fully working in the GUI albeit perhaps not the prettiest.

Screenshot 2021-05-07 at 13 10 32

**Comment by SeBBBe** [comment]
Seems like the current way of creating a default is not acceptable:

```
"hooks": {
        "preexecute": {
            "url": "",
            "method": "get",
            "payload": "",
            "respcodes": [
                200,
                201
            ],
            "contenttype": "application/json",
            "extractors": [],
            "headers": null
        },
        "postexecute": {
            "url": "",
            "method": "get",
            "payload": "",
            "respcodes": [
                200,
                201
            ],
            "contenttype": "application/json",
            "extractors": [],
            "headers": null
        }
    },
```

**Comment by DnlLrssn** [comment]
@SeBBBe hook headers should now show up in GUI better I hope?

**Comment by SeBBBe** [comment]
> @SeBBBe hook headers should now show up in GUI better I hope?

It works perfectly in the GUI! Will open a PR with the changes I already made to sort them out as a separate expander.

**Pull request merged**